packages/alfa-map/src/map.ts

```diff
@@ -48,59 +48,63 @@ export class Map<K, V> implements Collection.Keyed<K, V> {
     return this._size === 0;
   }
 
-  public forEach(callback: Callback<V, void, [K]>): void {
+  public forEach(callback: Callback<V, void, [key: K]>): void {
     Iterable.forEach(this, ([key, value]) => callback(value, key));
   }
 
-  public map<U>(mapper: Mapper<V, U, [K]>): Map<K, U> {
+  public map<U>(mapper: Mapper<V, U, [key: K]>): Map<K, U> {
     return new Map(this._root.map(mapper), this._size);
   }
 
-  public flatMap<L, U>(mapper: Mapper<V, Map<L, U>, [K]>): Map<L, U> {
+  public apply<U>(mapper: Map<K, Mapper<V, U>>): Map<K, U> {
+    return mapper.flatMap((mapper) => this.map(mapper));
+  }
```

Comment:
I'm not sure I understand what it means to apply a map to another map.
Can we add some tests of that?

Trying
```javascript
const map = Map.of(["foo", 1], ["bar", 2], ["baz", 3], ["qux", 4]);
  const fmap = Map.of(
    ["a", (x: number) => x + 1],
    ["b", (x: number) => x + 2],
    ["c", (x: number) => x + 3],
    ["d", (x: number) => x + 4]
  );
```
gives `[ [ 'baz', 7 ], [ 'qux', 8 ], [ 'foo', 5 ], [ 'bar', 6 ] ]` (so only the fourth function was applied?)
But with
```javascript
const fmap = Map.of(
    ["foo", (x: number) => x + 1],
    ["bar", (x: number) => x + 2],
    ["baz", (x: number) => x + 3],
    ["qux", (x: number) => x + 4]
  );
```
I get `[ [ 'baz', 5 ], [ 'qux', 6 ], [ 'foo', 3 ], [ 'bar', 4 ] ]` (so it's another function that get applied).
And I'm not sure I understand the differences nor the impact that similar keys has (or doesn't have).


---
Intuitively, I would have said that if I get a M={ (k, v), k in K and v in V } and a F={ (k, f), k in K and f in V -> U} then applying F to M should give { (k, f(v)) / (k,v) in M and (k, f(v)) in F} but that is clearly not what running the tests do…

---
What I guess happens is that each f is applied to each v; then the resulting f(v) is associated with the key that was associated to v in the first place; so for a given (k, v) we now have several (k, f₁(v)), (k, f₂(v)), …; and then only one of those "wins" and stays in the resulting map but the key collision resolution depends on the keys of the f in a somewhat incontrollable way 🤔

Reply:
Usually, `#apply()` will give you the Cartesian product between a collection of functions and a collection of values. For non-keyed types, this makes sense:

```
[1, 2].apply([x => x + 1, x => x * 2]) == [1 + 1, 2 + 1, 1 * 2, 2 * 2]
```

For keyed types like a map, however, I guess it doesn't really sense 🤔 Let's try:

```
[["a", 1], ["b", 2]].apply([["f", x => x + 1], ["g", x => x * 2]]) == [["a", 1 + 1], ["b", 2 + 1], ["a", 1 * 2], ["b", 2 * 2]]
```

Unless we're dealing with a multi map, this will result in keys colliding and the final result will either be `[["a", 1 + 1], ["b", 2 + 1]]` _or_ `[["a", 1 * 2], ["b", 2 * 2]]`.

Can we think of a more appropriate implementation of `#apply()` for keyed types?

Reply:
One option is performing pairwise application similar to Haskell's [`ZipList`](https://hackage.haskell.org/package/base-4.15.0.0/docs/Control-Applicative.html#t:ZipList):

```js
[["a", 1], ["b", 2]].apply([["a", x => x + 1], ["b", x => x * 2]]) == [["a", 1 + 1], ["b", 2 * 2]]
```

Reply:
Yes, the cartesian product makes sense for non-keyed collections. But for keyed collection we somehow need to combine keys or handle collisions in a somewhat intuitive way.

---
Getting a bit away from the code and the typing, on the mathematical side I would see 1 "natural" way of doing cartesian product of keyed values (i.e sets of pairs):
The pure cartesian product of `[ ["a", 1], ["b", 2]] × [[X, f], [Y, g]]` would be `[ ["a", 1, X, f], ["a", 1, Y, g], ["b", 2, X, f], ["b", 2, Y, g] ]`. Now, we are in a case where we want `f` and `g` to be functions, so we know how to combine them and can get `[ ["a", f(1), X], ["a", g(1), Y], ["b", f(2), X], ["b", g(2), Y] ]`. But from there, we need a way to somehow combine the keys, these could be:
1. Drop the `X/Y` key, only keep the `"a"/"b"` one and rely on some key collision resolution. This is the current implementation. I feel (as a dev) the collision resolution is too obscur for me to rely on it (as in, I'll never use this cartesian product so maybe we should just not have it).
2. Drop the `"a"/"b"` key and keep the `X/Y` one. Same problem.
3. Use an implicit but clear collision solving, typically only keep values where both keys are "equal".
4. Require an explicit way to combine keys, that is an extra parameter of type `comb: (K, K) => L` and then build `[ [comb("a", X), f(1)], [comb("a", Y), g(1)], [comb("b", X), f(2)], [comb("b", Y), g(2)] ]`. Collisions may still happen but are now the problem of the caller. Might be better to have `comb: (K, K) => Option<L>` and we drop the combinations that return `None` (so that 3. is just `comb = (k1, k2) => Equatable.equals(k1, k2) ? Some.of(k1) : None`, and `comb` might be optional defaulting to this). This may cause problem with typing, however…
5. Require that the keys of the second map are of type `K => L` (or `K => Option<L>`) so we can combine keys the same way we combine values: `[ [X("a"), f(1)], [Y("a"), g(1)], [X("b"), f(2)], [Y("b"), g(2)] ]`. That sounds like asking a bit too much and I'm not sure what would the use case be… (plus maybe also problematic type-wise). Again, possible collisions are now the problem of the caller.

I'm not sure which one makes more sense… It also depends on what use case we envision for that. I do feel that case 3. is the most natural cartesian product of keyed values (cartesian product of values whose key match). I like the added flexibility of 4. if it's possible…